From 92ab0e869ebb564d323b1814a818fd88f3b2d2ca Mon Sep 17 00:00:00 2001
From: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Date: Sun, 5 Feb 2023 18:46:32 +0200
Subject: [PATCH 1/2] build: add initial support for building

Enable the project building and running in non-interactive mode
with BSD-3-Clause licensed library libedit instead of GPLv3+ licensed
library readline.

Upstream-Status: Pending
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
---

 Makefile.am                   |  3 ++-
 Makefile.tools                | 18 +++++++++---------
 attrib/interactive.c          | 10 ++++++++++
 client/display.c              | 15 +++++++++++++++
 configure.ac                  |  2 ++
 src/shared/shell.c            | 26 ++++++++++++++++++++++++++
 tools/mesh-gatt/onoff-model.c |  6 ++++++
 tools/mesh-gatt/prov.c        |  5 +++++
 tools/obex-client-tool.c      |  4 ++++
 9 files changed, 79 insertions(+), 10 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 0fd405e8c..89519acc4 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -24,7 +24,8 @@ pkgincludedir = $(includedir)/bluetooth
 
 pkginclude_HEADERS =
 
-AM_CFLAGS = $(WARNING_CFLAGS) $(MISC_CFLAGS) $(UDEV_CFLAGS) $(LIBEBOOK_CFLAGS) $(LIBEDATASERVER_CFLAGS) $(ell_cflags)
+AM_CFLAGS = $(MISC_CFLAGS) $(WARNING_CFLAGS) $(UDEV_CFLAGS) $(LIBEBOOK_CFLAGS) \
+				$(LIBEDATASERVER_CFLAGS) $(READLINE_CFLAGS) $(ell_cflags)
 AM_LDFLAGS = $(MISC_LDFLAGS)
 
 if DATAFILES
diff --git a/Makefile.tools b/Makefile.tools
index 8cfaf8972..efe4fcdb0 100644
--- a/Makefile.tools
+++ b/Makefile.tools
@@ -13,7 +13,7 @@ client_bluetoothctl_SOURCES = client/main.c \
 					client/admin.h \
 					client/admin.c
 client_bluetoothctl_LDADD = gdbus/libgdbus-internal.la src/libshared-glib.la \
-				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline
+				$(GLIB_LIBS) $(DBUS_LIBS) $(READLINE_LIBS)
 endif
 
 if ZSH_COMPLETIONS
@@ -326,7 +326,7 @@ tools_meshctl_SOURCES = tools/meshctl.c \
 				tools/mesh-gatt/onoff-model.c
 tools_meshctl_LDADD = gdbus/libgdbus-internal.la src/libshared-glib.la \
 				lib/libbluetooth-internal.la \
-				$(GLIB_LIBS) $(DBUS_LIBS) -ljson-c -lreadline
+				$(GLIB_LIBS) $(DBUS_LIBS) -ljson-c $(READLINE_LIBS)
 
 bin_PROGRAMS +=  tools/mesh-cfgclient
 
@@ -342,7 +342,7 @@ tools_mesh_cfgclient_SOURCES = tools/mesh-cfgclient.c \
 				mesh/crypto.h mesh/crypto.c
 
 tools_mesh_cfgclient_LDADD = lib/libbluetooth-internal.la src/libshared-ell.la \
-						$(ell_ldadd) -ljson-c -lreadline
+				$(ell_ldadd) -ljson-c $(READLINE_LIBS)
 
 bin_PROGRAMS +=  tools/mesh-cfgtest
 
@@ -440,7 +440,7 @@ noinst_PROGRAMS += tools/btmgmt tools/obex-client-tool tools/obex-server-tool \
 tools_obex_client_tool_SOURCES = $(gobex_sources) $(btio_sources) \
 						tools/obex-client-tool.c
 tools_obex_client_tool_LDADD = lib/libbluetooth-internal.la \
-						$(GLIB_LIBS) -lreadline
+				$(GLIB_LIBS) $(READLINE_LIBS)
 
 tools_obex_server_tool_SOURCES = $(gobex_sources) $(btio_sources) \
 						tools/obex-server-tool.c
@@ -449,15 +449,15 @@ tools_obex_server_tool_LDADD = lib/libbluetooth-internal.la $(GLIB_LIBS)
 tools_bluetooth_player_SOURCES = tools/bluetooth-player.c
 tools_bluetooth_player_LDADD = gdbus/libgdbus-internal.la \
 				src/libshared-glib.la \
-				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline
+				$(GLIB_LIBS) $(DBUS_LIBS) $(READLINE_LIBS)
 
 tools_obexctl_SOURCES = tools/obexctl.c
 tools_obexctl_LDADD = gdbus/libgdbus-internal.la src/libshared-glib.la \
-				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline
+				$(GLIB_LIBS) $(DBUS_LIBS) $(READLINE_LIBS)
 
 tools_btmgmt_SOURCES = tools/btmgmt.c src/uuid-helper.c client/display.c
 tools_btmgmt_LDADD = lib/libbluetooth-internal.la src/libshared-mainloop.la \
-				-lreadline
+				$(READLINE_LIBS)
 if DEPRECATED
 noinst_PROGRAMS += attrib/gatttool
 
@@ -467,7 +467,7 @@ attrib_gatttool_SOURCES = attrib/gatttool.c attrib/att.c attrib/gatt.c \
 				attrib/utils.c src/log.c client/display.c \
 				client/display.h
 attrib_gatttool_LDADD = lib/libbluetooth-internal.la \
-			src/libshared-glib.la $(GLIB_LIBS) -lreadline
+				src/libshared-glib.la $(GLIB_LIBS) $(READLINE_LIBS)
 
 endif
 endif
@@ -513,5 +513,5 @@ tools/btpclient.$(OBJEXT): src/libshared-ell.la ell/internal
 
 tools_btpclientctl_SOURCES = tools/btpclientctl.c client/display.c
 tools_btpclientctl_LDADD = src/libshared-mainloop.la src/libshared-glib.la \
-				lib/libbluetooth-internal.la -lreadline
+				lib/libbluetooth-internal.la $(READLINE_LIBS)
 endif
diff --git a/attrib/interactive.c b/attrib/interactive.c
index 171b95738..eb0b4d6b4 100644
--- a/attrib/interactive.c
+++ b/attrib/interactive.c
@@ -23,8 +23,12 @@
 #include <sys/signalfd.h>
 #include <glib.h>
 
+#ifdef HAVE_EDITLINE_READLINE
+#include <editline/readline.h>
+#else
 #include <readline/readline.h>
 #include <readline/history.h>
+#endif
 
 #include "lib/bluetooth.h"
 #include "lib/sdp.h"
@@ -922,15 +926,19 @@ static gboolean signal_handler(GIOChannel *channel, GIOCondition condition,
 
 	switch (si.ssi_signo) {
 	case SIGINT:
+#ifndef HAVE_EDITLINE_READLINE
 		rl_replace_line("", 0);
 		rl_crlf();
+#endif
 		rl_on_new_line();
 		rl_redisplay();
 		break;
 	case SIGTERM:
 		if (__terminated == 0) {
+#ifndef HAVE_EDITLINE_READLINE
 			rl_replace_line("", 0);
 			rl_crlf();
+#endif
 			g_main_loop_quit(event_loop);
 		}
 
@@ -999,7 +1007,9 @@ int interactive(const char *src, const char *dst,
 	signal = setup_signalfd();
 
 	rl_attempted_completion_function = commands_completion;
+#ifndef HAVE_EDITLINE_READLINE
 	rl_erase_empty_line = 1;
+#endif
 	rl_callback_handler_install(get_prompt(), parse_line);
 
 	g_main_loop_run(event_loop);
diff --git a/client/display.c b/client/display.c
index a590a1388..05930e48d 100644
--- a/client/display.c
+++ b/client/display.c
@@ -18,7 +18,13 @@
 #include <stdarg.h>
 #include <stdbool.h>
 #include <ctype.h>
+
+#ifdef HAVE_EDITLINE_READLINE
+#include <string.h>
+#include <editline/readline.h>
+#else
 #include <readline/readline.h>
+#endif
 
 #include "display.h"
 
@@ -30,6 +36,8 @@ static void *saved_user_data = NULL;
 void rl_printf(const char *fmt, ...)
 {
 	va_list args;
+
+#ifndef HAVE_EDITLINE_READLINE
 	bool save_input;
 	char *saved_line;
 	int saved_point;
@@ -43,11 +51,13 @@ void rl_printf(const char *fmt, ...)
 		rl_replace_line("", 0);
 		rl_redisplay();
 	}
+#endif
 
 	va_start(args, fmt);
 	vprintf(fmt, args);
 	va_end(args);
 
+#ifndef HAVE_EDITLINE_READLINE
 	if (save_input) {
 		rl_restore_prompt();
 		rl_replace_line(saved_line, 0);
@@ -55,6 +65,7 @@ void rl_printf(const char *fmt, ...)
 		rl_forced_update_display();
 		free(saved_line);
 	}
+#endif
 }
 
 void rl_hexdump(const unsigned char *buf, size_t len)
@@ -121,7 +132,9 @@ void rl_prompt_input(const char *label, const char *msg,
 								label, msg);
 	rl_set_prompt(prompt);
 
+#ifndef HAVE_EDITLINE_READLINE
 	rl_replace_line("", 0);
+#endif
 	rl_redisplay();
 }
 
@@ -136,7 +149,9 @@ int rl_release_prompt(const char *input)
 	/* This will cause rl_expand_prompt to re-run over the last prompt, but
 	 * our prompt doesn't expand anyway. */
 	rl_set_prompt(saved_prompt);
+#ifndef HAVE_EDITLINE_READLINE
 	rl_replace_line("", 0);
+#endif
 	rl_point = saved_point;
 	rl_redisplay();
 
diff --git a/configure.ac b/configure.ac
index 5653e05d1..5ebe9b548 100644
--- a/configure.ac
+++ b/configure.ac
@@ -283,8 +283,10 @@ AM_CONDITIONAL(CLIENT, test "${enable_client}" != "no")
 if (test "${enable_client}" != "no" || test "${enable_mesh}" = "yes"); then
         AC_CHECK_HEADERS(readline/readline.h, enable_readline=yes,
                 AC_MSG_ERROR(readline header files are required))
+        READLINE_LIBS=-lreadline
 fi
 AM_CONDITIONAL(READLINE, test "${enable_readline}" = "yes")
+AC_SUBST(READLINE_LIBS)
 
 AC_ARG_ENABLE(systemd, AC_HELP_STRING([--disable-systemd],
 		[disable systemd integration]), [enable_systemd=${enableval}])
diff --git a/src/shared/shell.c b/src/shared/shell.c
index 21e8b50cd..eefd1f9a1 100644
--- a/src/shared/shell.c
+++ b/src/shared/shell.c
@@ -25,8 +25,12 @@
 #include <wordexp.h>
 #include <getopt.h>
 
+#ifdef HAVE_EDITLINE_READLINE
+#include <editline/readline.h>
+#else
 #include <readline/readline.h>
 #include <readline/history.h>
+#endif
 
 #include "src/shared/mainloop.h"
 #include "src/shared/timeout.h"
@@ -523,9 +527,11 @@ static int shell_exec(int argc, char *argv[])
 void bt_shell_printf(const char *fmt, ...)
 {
 	va_list args;
+#ifndef HAVE_EDITLINE_READLINE
 	bool save_input;
 	char *saved_line;
 	int saved_point;
+#endif
 
 	if (!data.input)
 		return;
@@ -537,6 +543,7 @@ void bt_shell_printf(const char *fmt, ...)
 		return;
 	}
 
+#ifndef HAVE_EDITLINE_READLINE
 	save_input = !RL_ISSTATE(RL_STATE_DONE);
 
 	if (save_input) {
@@ -547,6 +554,7 @@ void bt_shell_printf(const char *fmt, ...)
 		rl_replace_line("", 0);
 		rl_redisplay();
 	}
+#endif
 
 	va_start(args, fmt);
 	vprintf(fmt, args);
@@ -558,6 +566,7 @@ void bt_shell_printf(const char *fmt, ...)
 		va_end(args);
 	}
 
+#ifndef HAVE_EDITLINE_READLINE
 	if (save_input) {
 		if (!data.saved_prompt)
 			rl_restore_prompt();
@@ -566,6 +575,7 @@ void bt_shell_printf(const char *fmt, ...)
 		rl_forced_update_display();
 		free(saved_line);
 	}
+#endif
 }
 
 static void print_string(const char *str, void *user_data)
@@ -594,7 +604,9 @@ static void prompt_input(const char *str, bt_shell_prompt_input_func func,
 	data.saved_func = func;
 	data.saved_user_data = user_data;
 
+#ifndef HAVE_EDITLINE_READLINE
 	rl_save_prompt();
+#endif
 	bt_shell_set_prompt(str);
 }
 
@@ -654,7 +666,9 @@ int bt_shell_release_prompt(const char *input)
 
 	data.saved_prompt = false;
 
+#ifndef HAVE_EDITLINE_READLINE
 	rl_restore_prompt();
+#endif
 
 	func = data.saved_func;
 	user_data = data.saved_user_data;
@@ -684,7 +698,9 @@ static void rl_handler(char *input)
 	if (!input) {
 		rl_insert_text("quit");
 		rl_redisplay();
+#ifndef HAVE_EDITLINE_READLINE
 		rl_crlf();
+#endif
 		mainloop_quit();
 		return;
 	}
@@ -891,7 +907,9 @@ static char **menu_completion(const struct bt_shell_menu_entry *entry,
 			break;
 		}
 
+#ifndef HAVE_EDITLINE_READLINE
 		rl_completion_display_matches_hook = entry->disp;
+#endif
 		matches = rl_completion_matches(text, entry->gen);
 		break;
 	}
@@ -945,8 +963,10 @@ static void signal_callback(int signum, void *user_data)
 	switch (signum) {
 	case SIGINT:
 		if (data.input && !data.mode) {
+#ifndef HAVE_EDITLINE_READLINE
 			rl_replace_line("", 0);
 			rl_crlf();
+#endif
 			rl_on_new_line();
 			rl_redisplay();
 			return;
@@ -963,8 +983,10 @@ static void signal_callback(int signum, void *user_data)
 	case SIGTERM:
 		if (!terminated) {
 			if (!data.mode) {
+#ifndef HAVE_EDITLINE_READLINE
 				rl_replace_line("", 0);
 				rl_crlf();
+#endif
 			}
 			mainloop_quit();
 		}
@@ -1027,7 +1049,9 @@ static void rl_init(void)
 	setlinebuf(stdout);
 	rl_attempted_completion_function = shell_completion;
 
+#ifndef HAVE_EDITLINE_READLINE
 	rl_erase_empty_line = 1;
+#endif
 	rl_callback_handler_install(NULL, rl_handler);
 
 	rl_init_history();
@@ -1156,7 +1180,9 @@ static void rl_cleanup(void)
 	if (data.history[0] != '\0')
 		write_history(data.history);
 
+#ifndef HAVE_EDITLINE_READLINE
 	rl_message("");
+#endif
 	rl_callback_handler_remove();
 }
 
diff --git a/tools/mesh-gatt/onoff-model.c b/tools/mesh-gatt/onoff-model.c
index 9dfedd85a..46350a3d6 100644
--- a/tools/mesh-gatt/onoff-model.c
+++ b/tools/mesh-gatt/onoff-model.c
@@ -21,8 +21,14 @@
 #include <stdbool.h>
 #include <sys/uio.h>
 #include <wordexp.h>
+
+#ifdef HAVE_EDITLINE_READLINE
+#include <editline/readline.h>
+#else
 #include <readline/readline.h>
 #include <readline/history.h>
+#endif
+
 #include <glib.h>
 
 #include "src/shared/shell.h"
diff --git a/tools/mesh-gatt/prov.c b/tools/mesh-gatt/prov.c
index 1f3f62dfd..796de33ef 100644
--- a/tools/mesh-gatt/prov.c
+++ b/tools/mesh-gatt/prov.c
@@ -19,8 +19,13 @@
 #include <sys/uio.h>
 #include <wordexp.h>
 
+#ifdef HAVE_EDITLINE_READLINE
+#include <editline/readline.h>
+#else
 #include <readline/readline.h>
 #include <readline/history.h>
+#endif
+
 #include <glib.h>
 
 #include "src/shared/util.h"
diff --git a/tools/obex-client-tool.c b/tools/obex-client-tool.c
index ab9332896..2d4f50494 100644
--- a/tools/obex-client-tool.c
+++ b/tools/obex-client-tool.c
@@ -22,8 +22,12 @@
 #include <errno.h>
 #include <stdio.h>
 
+#ifdef HAVE_EDITLINE_READLINE
+#include <editline/readline.h>
+#else
 #include <readline/readline.h>
 #include <readline/history.h>
+#endif
 
 #include "gobex/gobex.h"
 #include "btio/btio.h"
-- 
2.39.1

